### PR TITLE
create test and implement function for PCA

### PR DESCRIPTION
**Pull request by hankunxiao**

### CODE REVIEW

File: src/sharpedge/pca_compression.py [hunk 1]
```diff
@@ -1,44 +1,88 @@
...
+    if not isinstance(preservation_rate, (int, float)):
+        raise TypeError("preservation_rate must be a number.")
+    if not (0 < preservation_rate <= 1):
+        raise ValueError("preservation_rate must be a float between 0 and 1.")
+    if preservation_rate < 0.1:
+        print("Warning: Very low preservation_rate may result in significant quality loss.")
+    
+    # Raise a warning for low preservation_rate
+    if preservation_rate < 0.1:
+        warnings.warn("Very low preservation_rate may result in significant quality loss.", UserWarning)
```
Comment (Arc-Celt) [inline src/sharpedge/pca_compression.py:59]: I noticed that the code contains both a print statement and a warning for the same condition (preservation_rate < 0.1), which might seem a bit redundant. Would it make better sense to retain only the warning and remove the print statement?

File: tests/test_pca_compression.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import pytest
+import numpy as np
+from sharpedge.pca_compression import pca_compression
+
+# Expected Test Cases
+@pytest.mark.parametrize("input_img, preservation_rate, expected_output", [
+    (np.array([[25, 15], [15, 25]]), 0.5, np.array([[20, 20], [20, 20]])),  # Reduce by 50%    
+    (np.array([[25, 15], [15, 25]]), 1.0, np.array([[25, 15], [15, 25]])),  # Reduce by 0%
```
Comment (Arc-Celt) [inline tests/test_pca_compression.py:8]: I suggest adding two additional expected test cases to cover the following scenarios:

1. An array containing larger numbers.

1. An array with more elements (larger array size).

Here are two test cases I’ve come up with for your reference:
```
(np.array([[100, 200], [200, 100]]), 0.75, np.array([[150, 150], [150, 150]])),  # Reduce by 25%
(np.array([[50, 150, 200], [120, 180, 160], [80, 160, 140]]), 0.2, np.array([[ 85., 165., 168.], [ 90., 176., 179.], [ 77., 149., 152.]])),  # Reduce by 80%
```

File: tests/test_pca_compression.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        compressed_img, expected_output, decimal=0,
+        err_msg=f"Output mismatch for input {input_img} with preservation_rate={preservation_rate}"
+    )
+
+# Edge Cases: 
+def test_low_preservation_rate_warning():
+    """Test that a warning is raised for very low preservation_rate (< 0.1)."""
+    img = np.ones((100, 100))
+    with pytest.warns(UserWarning, match="Very low preservation_rate may result in significant quality loss."):
+        pca_compression(img, preservation_rate=0.05)
```
Comment (Arc-Celt) [inline tests/test_pca_compression.py:25]: For the edge cases, though it is just a rather simple check, you can add different cases that use different low preservation rates. FYR:

```
@pytest.mark.parametrize("img, preservation_rate, warning_message", [
    (np.ones((100, 100)), 0.05, "Very low preservation_rate may result in significant quality loss."),
    (np.ones((99, 101)), 0.01, "Very low preservation_rate may result in significant quality loss."),
    (np.ones((98, 102)), 0.09, "Very low preservation_rate may result in significant quality loss."),
])
```

Additionally, it might be a good idea to test the case where the preservation rate is exactly 0.1. Since this is the threshold set in the `pca_compression` function, where no warning is expected, testing this edge case can help ensure the function behaves as intended at this boundary. While it's not necessary, it could be a useful check.

File: tests/test_pca_compression.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    """Test that a negative preservation_rate raises a ValueError."""
+    img = np.ones((100, 100))
+    with pytest.raises(ValueError, match="preservation_rate must be a float between 0 and 1."):
+        pca_compression(img, preservation_rate=-0.1)
+
+def test_invalid_input_not_2d():
+    """Test that a non-2D input raises a ValueError."""
+    img = np.ones((100, 100, 3))
+    with pytest.raises(ValueError, match="Input image must be a 2D array."):
+        pca_compression(img)
```
Comment (Arc-Celt) [inline tests/test_pca_compression.py:44]: How about adding more invalid preservation rate types with @pytest.mark.parametrize? (just to make our test more comprehensive) list, tuples, numpy arrays, all sorts of dtypes.

Also, apart from when the preservation_rate is negative, you could try cases where preservation_rate is larger than 1.

**Review by Arc-Celt — COMMENTED** [review]
Great job on the complex compression function! As per the Milestone requirements, unit tests should cover 3 to 5 cases for each of the 3 categories (expected, edge, and error cases), so it might be helpful to add a few user cases to further enrich your tests.

In my comments, I've recommended a few additional test cases to further ensure comprehensive coverage. Some of them might not be necessary though. Feel free to decide if you want to include them!

**Comment by hankunxiao** [comment]
@Arc-Celt thanks for you feedback, I have made the changes.

**Review by Arc-Celt — APPROVED** [review]
Everything looks great! And it achieves a coverage of 100% during test. Should be ok to merge now.
